### PR TITLE
DAOS-16170 control: Ignore EngineDied event for old incarnation

### PR DESCRIPTION
**Pull request by kjacque**
It is possible to be forwarded an EngineDied event late, after the engine has re-joined. This can incorrectly re-mark the rank as Errored.

- Include incarnation in engine-related events.
- Print incarnation in logs if provided.
- Do not update member if engine died event is for old incarnation.

Features: control

### Steps for the author:

* [ ] Commit message follows the [guidelines](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Appropriate [Features or Test-tag](https://daosio.atlassian.net/wiki/spaces/DC/pages/10984259629/Test+Tags) pragmas were used.
* [ ] Appropriate [Functional Test Stages](https://daosio.atlassian.net/wiki/spaces/DC/pages/12147556353/CI+Functional+Test+Stages) were run.
* [ ] At least two positive code reviews including at least one code owner from each category referenced in the PR.
* [ ] Testing is complete. If necessary, forced-landing label added and a reason added in a comment.

#### After all prior steps are complete:
* [ ] Gatekeeper requested (daos-gatekeeper added as a reviewer).

### CODE REVIEW

**Comment by daosbuild3** [comment]
Test stage Functional Hardware Medium MD on SSD completed with status UNSTABLE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net/job/daos-stack/job/daos//view/change-requests/job/PR-16511/1/testReport/

**Comment by daosbuild3** [comment]
Test stage Functional Hardware Medium Verbs Provider MD on SSD completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-16511/1/execution/node/1427/log

**Comment by phender** [comment]
@kjacque this PR needs to run with the `recovery` tag/feature.

**Comment by kjacque** [comment]
> @kjacque this PR needs to run with the `recovery` tag/feature.

Good catch. I rebased on master and amended the commit pragma.

**Comment by daosbuild3** [comment]
Test stage Functional Hardware Medium MD on SSD completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-16511/3/execution/node/627/log

**Comment by daosbuild3** [comment]
Test stage Functional Hardware Medium Verbs Provider MD on SSD completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-16511/3/execution/node/672/log

**Comment by daosbuild3** [comment]
Test stage Functional Hardware Large MD on SSD completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-16511/5/execution/node/1338/log

File: src/control/system/membership.go [hunk 1]
```diff
@@ -711,6 +711,11 @@ func (m *Membership) handleEngineFailure(evt *events.RASEvent) {
 		return
 	}
 
+	if evt.Incarnation != 0 && evt.Incarnation < member.Incarnation {
```
Comment (tanabarr) [inline src/control/system/membership.go:714]: when would incarnation be zero and why should we skip this clause if it is?

**Review by tanabarr — APPROVED** [review]
looks like a good change

File: src/control/system/membership.go (hunk 1 above)
Comment (kjacque) [reply src/control/system/membership.go:714]: If the engine did not successfully join before dying, it's possible that the control plane never got an incarnation for the engine, as it's generated by the engine itself. In that case we don't want to skip processing the EngineDied event.

Comment (kjacque) [reply src/control/system/membership.go:714]: As I think about it more, I think we should record the incarnation ASAP after getting the "ready" response from the engine. Then we could probably avoid that case altogether.

Comment (liw) [inline src/control/system/membership.go:714]: It's good that we use `<` instead of `==` here, since there's subtlety with our incarnation numbers today:
```
Engine X joins with incarnation 1
Engine X bumps its incarnation from 1 to 2 because it finds that someone suspects it (The control plane is not informed directly about this.)
Engine X terminates
MS sees events for engine X with incarnation 2
```

**Review by liw — DISMISSED** [review]

File: src/control/system/membership.go (hunk 1 above)
Comment (kjacque) [reply src/control/system/membership.go:714]: Interesting! That's good to know, thanks.

Comment (kjacque) [reply src/control/system/membership.go:714]: Fixed with the latest commit. Thanks!

**Review by liw — DISMISSED** [review]

File: src/control/system/membership.go (hunk 1 above)
Comment (tanabarr) [inline src/control/system/membership.go:714]: with your most recent change can we not remove the zero-check here?

Comment (kjacque) [reply src/control/system/membership.go:714]: We could -- I'd like to leave it in case there's still some gap somewhere. Logically there shouldn't be, but it also makes no sense to check incarnation at all if it's unset. 0 is never valid.

**Comment by daosbuild3** [comment]
Test stage Functional Hardware Large MD on SSD completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-16511/7/execution/node/1368/log

**Comment by daosbuild3** [comment]
Test stage Functional Hardware Medium Verbs Provider MD on SSD completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-16511/7/execution/node/1323/log

File: src/control/system/membership.go (hunk 1 above)
Comment (tanabarr) [reply src/control/system/membership.go:714]: Shouldn't we be returning a failure if it's not set? otherwise looks like we are masking an invalid state.
I understand the scenario where we don't want to create a failure case but covering up undesired scenarios will also bite us potentially.

Comment (kjacque) [reply src/control/system/membership.go:714]: Understood... We actually don't return errors from this function. Maybe we should log it, though.

Comment (kjacque) [reply src/control/system/membership.go:714]: Spent a little more time looking into this. There actually is a case where 0 would not be surprising. Consider this case:
- Admin starts daos_server.
- daos_engine process starts and dies before it can send a dRPC to the control plane including its incarnation.
- The RAS event includes incarnation = 0, because the control plane never got an incarnation.

This would only happen immediately after starting up the daos_server. We could conceivably have an incarnation in the member DB at that point, if the engine had run successfully in the past, but we would not have a current incarnation. That is definitely a corner case, and could leave an opening for the same bug to come through (essentially: engine fails to come up, event is sent, engine successfully comes up, forwarded event for incarnation 0 arrives, engine is incorrectly marked as errored).

Comment (kjacque) [reply src/control/system/membership.go:714]: I spent a little time talking to @mjmac about this yesterday. I think I have a solution for this particular corner case and am working on it.

**Comment by daosbuild3** [comment]
Test stage Functional on EL 8.8 completed with status UNSTABLE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net/job/daos-stack/job/daos//view/change-requests/job/PR-16511/9/testReport/

**Review by tanabarr — APPROVED** [review]

**Review by tanabarr — APPROVED** [review]
probably want to mention superblock related updates somewhere in the commit message

**Comment by kjacque** [comment]
> probably want to mention superblock related updates somewhere in the commit message

Good idea, I added a line to the description so it can be used as the commit message.

File: src/control/server/instance_exec.go [hunk 2]
```diff
@@ -64,6 +81,10 @@ func (ei *EngineInstance) start(ctx context.Context) (chan *engine.RunnerExitInf
 		ei.log.Errorf("instance %d: unable to log SCM storage stats: %s", ei.Index(), err)
 	}
 
+	if err := ei.initIncarnationFromSuperblock(); err != nil {
```
Comment (liw) [inline src/control/server/instance_exec.go:84]: Related to [this comment](https://github.com/daos-stack/daos/pull/16511#discussion_r2190481128), when an engine exits before sending its incarnation to the local daos_server, this incarnation of the engine must have not joined the system yet, so the RAS event (actually, what RAS event that is?) with incarnation 0 may be safely ignored?

Comment (kjacque) [reply src/control/server/instance_exec.go:84]: So this is the EngineDied event. If the engine has never joined (incarnation 0), I think we can ignore EngineDied, you're right. There is no Member to update. If the engine has joined in the past and there is an incarnation in the superblock, the event is useful because it updates the Member status. If we use the last known incarnation, that should be OK.

Comment (liw) [reply src/control/server/instance_exec.go:84]: Sorry, my question is assuming that we don't use the last known incarnation, that is, without the newly-added superblock field. Using the last known incarnation seems to be lying (even though it might be okay for the current code). And to do so we are adding a new persistent field---I'm wondering if that's going too far, and if we should just ignore such an event, or I must be missing the point?

If we can't ignore the event, there may be alternative approaches. For instance, could daos_server generate the incarnation and pass it to daos_engine? This way, we wouldn't need to store it persistently in the superblock, and the event would become accurate (i.e., the incarnation in the event would be faithful). Again, that's _if we can't ignore the event_.

Comment (kjacque) [reply src/control/server/instance_exec.go:84]: Ah, I see. The issue with ignoring the event is potentially the following situation:
1. DAOS has been up and running for a while.
2. Admin shuts down daos_servers.
3. Admin brings daos_servers back up.
4. daos_engine begins to start up and then crashes.

In this case the incarnation would be 0. If we ignore the event, the state of the affected Member will *not* become "Errored". I'm not sure what the state would be.

I wouldn't be against the daos_server being the one who defines the incarnation. That requires a bit more surgery though. That also feels a little stranger in the case you described previously, where the engine bumps its own incarnation because it's considered suspect, but the daos_server doesn't know about that change.

Comment (liw) [reply src/control/server/instance_exec.go:84]: I think the incarnation bumps upon suspicions are largely a separate "problem". (We could easily notify daos_server upon those if there were good reasons.)

If we ignore the event in that case, and the member ends up in "starting" state, things will be too confusing. (If the member ends up in "stopped"/"excluded" state, I'd consider the state acceptable.) So it seems that we can't ignore the event...

OK, at least it sounds like the concern about the addition of the "incarnation" superblock field is well understood. Sorry for delaying this PR.

**Review by liw — APPROVED** [review]

File: src/control/server/instance_exec.go (hunk 2 above)
Comment (kjacque) [reply src/control/server/instance_exec.go:84]: No worries. I appreciate the discussion and insights. Thanks!

**Review by knard38 — APPROVED** [review]

**Comment by daosbuild3** [comment]
Test stage Functional on EL 8.8 completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-16511/11/execution/node/1268/log

**Comment by daosbuild3** [comment]
Test stage Functional Hardware Large MD on SSD completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-16511/12/execution/node/793/log

**Comment by daosbuild3** [comment]
Test stage Functional Hardware Medium MD on SSD completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-16511/12/execution/node/804/log

**Comment by daosbuild3** [comment]
Test stage Functional Hardware Medium MD on SSD completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-16511/13/execution/node/1498/log

**Comment by daosbuild3** [comment]
Test stage Functional Hardware Medium Verbs Provider MD on SSD completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-16511/13/execution/node/1509/log

**Comment by daosbuild3** [comment]
Test stage Functional Hardware Large MD on SSD completed with status UNSTABLE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net/job/daos-stack/job/daos//view/change-requests/job/PR-16511/13/testReport/

**Comment by daosbuild3** [comment]
Test stage Functional Hardware Large MD on SSD completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-16511/14/execution/node/1464/log

**Comment by kjacque** [comment]
Test failures are known issues:
- https://daosio.atlassian.net/browse/DAOS-17888
- https://daosio.atlassian.net/browse/DAOS-17751

I know we can't land anything until CI is up next week, but whenever things re-open, this one is ready.
